### ISSUE TITLE
[aws-rds] ParameterGroup is missing from CloudFormation template

### ISSUE THREAD

**Issue by manospasj** [aws/aws-cdk#9741]
Since version 1.58.0 ParameterGroup resources are missing from the generated CloudFormation template.

### Reproduction Steps

- Create a new cdk app using version 1.51.0 and add the following code in the stack:
```
new ParameterGroup(this, "RDS-ParameterGroup", {
  family: "mysql:5.7",
});
```
- Run the `cdk synth` command and verify that the `AWS::RDS::DBParameterGroup` resource exists in the generated CloudFormation template.
- Now update cdk to version 1.58.0 and modify the ParameterGroup to:
```
new ParameterGroup(this, "RDS-ParameterGroup", {
  engine: DatabaseInstanceEngine.mysql({
    version: MysqlEngineVersion.VER_5_7,
  }),
});
```
- Run the `cdk synth` command and view the generated CloudFormation template. The `AWS::RDS::DBParameterGroup` resource is missing.

### What did you expect to happen?

The `AWS::RDS::DBParameterGroup` resource should exist in both versions of aws cdk.

### What actually happened?

The `AWS::RDS::DBParameterGroup` resource exists in version 1.51.0 but not in 1.58.0.

### Environment

  - **CLI Version:** 1.58.0
  - **Framework Version:** 1.58.0
  - **Node.js Version:** 14.5.0
  - **OS:** Windows 10
  - **Language (Version):** TypeScript (3.9.6)

--- 

This is :bug: Bug Report

**Comment by Chriscbr** [comment]
It looks like this was caused by [#8959](https://github.com/aws/aws-cdk/pull/8959), and the current behavior is that the parameter group only gets constructed when it gets bound to an RDS cluster or instance - I think this can be achieved by either passing it in as a parameter in your constructor, as shown in this example [here](https://github.com/aws/aws-cdk/blob/b1b9d23f442b674df97c1aadb4f84cc0251bb826/packages/%40aws-cdk/aws-rds/test/integ.instance.lit.ts#L68), or by manually calling the bindToInstance or bindToCluster methods described in the docs [here](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-rds.ParameterGroup.html).

Would either of these options work for your use case? If not, I think this would be a pretty easy fix to return to the previous behavior.

**Comment by manospasj** [comment]
I don't want to bound the parameter group to an instance, as the RDS instances are not part of my cdk app. Technically creating a parameter group without an RDS instance is possible using CloudFormation, I don't see why cdk should behave in a different way.

Also, since the bindToInstance doesn't require the instance as a parameter, how will this help? What does this method do?

**Comment by Chriscbr** [comment]
If you call the method like so, you should find that the parameter group now appears in the CloudFormation template:
```
new ParameterGroup(this, "RDS-ParameterGroup", {
  engine: DatabaseInstanceEngine.mysql({
    version: MysqlEngineVersion.VER_5_7,
  }),
}).bindToInstance({});
```
Maybe @skinny85 can shed some light here - but I think the reason it's implemented like this currently is that right now, a ParameterGroup can be constructed to make either a [DBParameterGroup](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-rds-dbparametergroup.html) by calling `bindToInstance`, or a [DBClusterParameterGroup](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-rds-dbclusterparametergroup.html) by calling `bindToCluster`, so the class is somewhat multi-purpose. But perhaps the docs could be clearer about this to explain how these work / which resources they generate.

(It also looks like right now, if you tried to call `bindToInstance()` and then `bindToCluster()` afterwards, it would just ignore the second call, and perhaps it should really throw an error - but that's a separate issue.)

**Comment by manospasj** [comment]
Thanks for your explanation, that makes perfect sense now. I agree that the docs could be better on this, as I spend quite some time reading them before opening this ticket and I would never guess that this is how it works.

Also, [this](https://github.com/aws/aws-cdk/blob/fab7e28f1cdae2eb65dddd32142fe64dc6955d63/packages/%40aws-cdk/aws-rds/lib/parameter-group.ts#L92) line in the source code needs to be updated as it suggests that the ParameterGroup class can only create a AWS::RDS::DBParameterGroup resource. The AWS::RDS::DBClusterParameterGroup resource needs to be included.

**Comment by skinny85** [comment]
Hey @manospasj ,

yes, the issue is exactly what @Chriscbr said. Basically, the `ParameterGroup` class is used both for `AWS::RDS::DBParameterGroup` and `AWS::RDS::DBClusterParameterGroup`, and creates the correct underlying resource when it's passed to `Instance` or `Cluster`, respectively.

> Also, [this](https://github.com/aws/aws-cdk/blob/fab7e28f1cdae2eb65dddd32142fe64dc6955d63/packages/%40aws-cdk/aws-rds/lib/parameter-group.ts#L92) line in the source code needs to be updated as it suggests that the ParameterGroup class can only create a AWS::RDS::DBParameterGroup resource. The AWS::RDS::DBClusterParameterGroup resource needs to be included.

Unfortunately, we can't have multiple CloudFormation resource types there. But I'll try to update the docs to make it sure it's more clear.

> (It also looks like right now, if you tried to call `bindToInstance()` and then `bindToCluster()` afterwards, it would just ignore the second call, and perhaps it should really throw an error - but that's a separate issue.)

Nope, that is not the behavior currently. The second call, to `bindToCluster()`, will actually create the `AWS::RDS::DBClusterParameterGroup`. So you'll have 2 groups, an `AWS::RDS::DBParameterGroup` and an `AWS::RDS::DBClusterParameterGroup`.

**Comment by lossanarch** [comment]
imho this bindTo...() method is weird and non-standard when compared to the rest of cdk. It may result in a small amount of duplication but would be more transparent and intuitive to reverse #8959 and just have the object constructed as normal and have the user decide which they want by deliberately constructing one or the other.
This is magic, which is why it's catching people out and needs better documentation. It would document itself if it were two separate classes. Just my 2c.

**Comment by skinny85** [comment]
@lossanarch I'm not disagreeing with your assessment, but it's a little late for that now 😛 (not sure how easy it would be to reverse https://github.com/aws/aws-cdk/pull/8959 without introducing a breaking change).

**Comment by lossanarch** [comment]
Yeah you're probably right, likely a fairly hard breaking change. Sorry, I only first came across this today.
I wonder if the bindTo methods could be left in a working but deprecated state to prevent the break, while re/adding the NewParameterGroup/NewClusterParameterGroup methods. Looks feasible at first glance although I haven't had an in-depth look yet.
Hey thanks for the quick response 👍

**Comment by gudipati** [comment]
Hi @manospasj,

The documentation now shows that you can create a parameter group for a cluster or instance - [aws-rds/README.md](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-rds/README.md?plain=1#L1267)

Does this solve your original issue ?

**Comment by manospasj** [comment]
Hello @gudipati,

Unfortunately I can't confirm or deny because I'm not working on this project any more.

I believe though that it's still an issue because the documentation mentions that the parameter group needs to be bound to an RDS instance, but this issue is when you want to create the parameter group and NOT bound it to an RDS instance.

**Comment by gudipati** [comment]
Hi @manospasj,

As per the comments above, the parameter group gets created only when it gets bound to an RDS cluster or instance and the documentation now reflects that - [aws-rds/README.md](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-rds/README.md?plain=1#L1267).

Could you explain your usecase for creating a ParameterGroup without attaching it to an RDS cluster or instance ? Is this feature still required ?

**Comment by rgoltz** [comment]
From an Ops/CloudFormation perspective, when performing an RDS upgrade, it's common to encounter a new default ParameterGroup that corresponds to the version you're upgrading to. By creating this new ParameterGroup first (I did this using plain CloudFormation templates so far), you can more easily compare the differences between the existing and new Parameter Groups. This approach allows you to understand any changes and helps in creating a custom ParameterGroup that matches the version you're upgrading to.

Once you've set up the new ParameterGroup, you'll need to reapply any customizations from the previous version. Additionally, this process can highlight new default settings in the upgraded major version (Version A+1) that didn't exist in the previous version (Version A). These changes may appear as though you've added or modified settings, when in fact, they're simply new defaults in the latest version.

**Comment by gudipati** [comment]
Hi @rgoltz,
Thank you for the detailed explanation of your use case.
If you're interested in contributing a fix, a PR would be most welcome!

**Comment by gudipati** [comment]
You can now create standalone parameter groups by following the [README](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-rds/README.md#creating-standalone-parameter-groups)
